### PR TITLE
Add OpenStreetMap lookups

### PR DESCRIPTION
**Pull request by gsvr**

### CODE REVIEW

**Comment by gsvr** [comment]
@imsickofmaps Please do a preliminary review.  I think I should still add more tests that check the different opts choices.

File: lib/location.js [hunk 1]
```diff
@@ -80,19 +161,25 @@ var LocationState = State.extend(function (self, name, opts) {
...
                         'Please try another search term.'
                         ].join('\n'),
+        mapping_service:"gmaps",
         options_per_page:8,
         characters_per_page:160,
         next_text:"Next",
         previous_text:"Previous",
-        store_fields:['formatted_address'],
-        namespace:'location'
+        store_fields: ((opts.mapping_service === 'osmaps') ?
```
Comment (imsickofmaps) [inline lib/location.js]: If your other default is gmaps I think the default here should be that too. Don't try and get too smart IMHO.

File: lib/location.js [hunk 2]
```diff
@@ -343,7 +522,27 @@ var LocationState = State.extend(function (self, name, opts) {
     self.display = function() {
         return self.metadata.display_text || self.question_text;
     };
-    
+
+    self.get_viewbox = function() {
+        if (self.country_code) {
```
Comment (imsickofmaps) [inline lib/location.js]: Take this logic out and make it so bounding_box is set at configuration time from the country_boxes values if that is the user pref. Removes the confusion as to precedence.

File: lib/location.js [hunk 3]
```diff
@@ -343,7 +522,27 @@ var LocationState = State.extend(function (self, name, opts) {
...
+            self.bounding_box = self.country_boxes[self.country_code];
+        }
+        return "viewbox=" + self.bounding_box[0] + "%2C" + self.bounding_box[1] +
```
Comment (imsickofmaps) [inline lib/location.js]: `self.bounding_box.join('%2C')` cleaner?

**Comment by hodgestar** [comment]
Very high-level comment -- I think having splitting out helpers for Google Maps and Openstreet Map and then just setting a ``self.map_api`` instance on the ``LocationState`` would make the implementation a lot cleaner?

**Comment by imsickofmaps** [comment]
@hodgestar LOL! We just had this exact conversation in the office...  :)

**Comment by hodgestar** [comment]
:D

File: lib/location.js [hunk 4]
```diff
@@ -91,14 +340,34 @@ var LocationState = State.extend(function (self, name, opts) {
...
+
+        mapping_service: "gmaps",
+        map_api_opts: {},
     });
 
     State.call(self, name, opts);
 
+    // Set mapping service
+    var map_api;
+    if (opts.mapping_service === 'gmaps') {
```
Comment (imsickofmaps) [inline lib/location.js]: Do we need this? Shouldn't it just default to gmaps unless it's over-written?

File: lib/testing.js [hunk 5]
```diff
@@ -45,33 +45,67 @@ module.exports = function() {
             request:"New Street",
             address_list:["New Street 1", "New Street 2"]
         });
-    
+
     */
     var add_location = function(opts) {
```
Comment (gsvr) [inline lib/testing.js]: Issue #21 opened for refactoring this to be more generic.

File: lib/location.js [hunk 6]
```diff
@@ -6,13 +6,203 @@ var State = states.State;
...
+            delete data.formatted_address;
+        }
+    }
+
+};
+
 var LocationState = State.extend(function (self, name, opts) {
     /**class:LocationState(name, opts)
 
     A state which requests a location from the user, and sets the user data
```
Comment (gsvr) [inline lib/location.js]: Docs to be updated.

**Comment by hodgestar** [comment]
Superseded by #23 and #24.

**Pull request closed without merging**